File: src/KPLinguaPreprocessing/Python/parser/KplIteratorBuilder.py
from antlr4 import InputStream, CommonTokenStream

from tools.kPLExtension.Models import Variable, Number, ArithmeticExpression, ArithmeticOperations, Iterator, Sign, \
    LogicalOperators, LogicalExpression, RelationalExpression, RelationalOperators, RelationalExpressionTrue
from tools.kPLExtension.parser.grammar.KplIteratorLexer import KplIteratorLexer
from tools.kPLExtension.parser.grammar.KplIteratorParser import KplIteratorParser
from tools.kPLExtension.parser.grammar.KplIteratorVisitor import KplIteratorVisitor


class KplIteratorBuilder(KplIteratorVisitor):

    def __init__(self, variables):
        self.variables = variables

    def visitIterators(self, ctx:KplIteratorParser.IteratorsContext):
        iterators = list(map(lambda a: self.visit(a), ctx.children))
        iterators = list(filter(lambda a: a is not None, iterators))
        iterators = list(filter(lambda a: isinstance(a, Iterator), iterators))
        restrictions = RelationalExpressionTrue()
        if ctx.restrictions is not None:
            restrictions = self.visit(ctx.restrictions)
        return iterators, restrictions

    def visitIterator(self, ctx:KplIteratorParser.IteratorContext):
        leftInq = Sign(ctx.leftInq.text)
        rightInq = Sign(ctx.rightInq.text)
        e1 = self.visit(ctx.left)
        e2 = self.visit(ctx.right)
        identifier = ctx.Identifier().getText()
        increment = 1
        if ctx.increment is not None:
            increment = int(ctx.increment.text)
        if identifier in self.variables:
            variable = self.variables[identifier]
        else:
            variable = Variable(identifier)
            self.variables[identifier] = variable
        return Iterator(variable, e1, leftInq, e2, rightInq, increment)

    def visitArOpExp(self, ctx:KplIteratorParser.ArOpExpContext):
        if ctx.op is not None:
            o = ctx.op.text
            e1 = self.visit(ctx.left)
            e2 = self.visit(ctx.right)
            return ArithmeticExpression(e1, ArithmeticOperations(o), e2)
        elif ctx.Number() is not None:
            return Number(int(ctx.getText()))
        elif ctx.identifierOp is not None:
            identifier = self.visit(ctx.identifierOp)
            return identifier

    def visitLoOpExp(self, ctx:KplIteratorParser.LoOpExpContext):
        if ctx.op is not None:
            op = RelationalOperators(ctx.op.text)
            e1 = self.visit(ctx.left)
            e2 = self.visit(ctx.right)
            return RelationalExpression(e1, op, e2)
        else:
            return self.visit(ctx.lo)

    def visitLogicalExpression(self, ctx:KplIteratorParser.LogicalExpressionContext):
        ex1 = self.visit(ctx.left)
        op = self.visit(ctx.lo)
        ex2 = self.visit(ctx.right)
        return LogicalExpression(ex1, op, ex2)

    def visitLogicalOperatorName(self, ctx:KplIteratorParser.LogicalOperatorNameContext):
        return LogicalOperators(ctx.getText())

    def visitIdentifier(self, ctx:KplIteratorParser.ArOpExpContext):
        identifier = ctx.Identifier().getText()
        if identifier in self.variables:
            variable = self.variables[identifier]
        else:
            variable = Variable(identifier)
            self.variables[identifier] = variable
        return variable

    def visitParameters(self, ctx:KplIteratorParser.ParametersContext):
        variables = list(map(lambda a: self.visit(a), ctx.children))
        variables = list(filter(lambda a: a is not None, variables))
        return variables

    def visitParameter(self, ctx:KplIteratorParser.ParameterContext):
        identifier = ctx.Identifier().getText()
        number = int(ctx.Number().getText()) if ctx.Number() is not None else 0
        return Variable(identifier, number)

    def buildParameters(self, text):
        data = InputStream(text)
        lexer = KplIteratorLexer(data)
        stream = CommonTokenStream(lexer=lexer)
        parser = KplIteratorParser(stream)
        tree = parser.parameters()
        variables = self.visitParameters(tree)
        return variables

    def buildExpression(self, text):
        data = InputStream(text)
        lexer = KplIteratorLexer(data)
        stream = CommonTokenStream(lexer=lexer)
        parser = KplIteratorParser(stream)
        tree = parser.arithmeticExpression()
        s = self.visitArOpExp(tree)
        return s

    def buildLogicalExpression(self, text):
        data = InputStream(text)
        lexer = KplIteratorLexer(data)
        stream = CommonTokenStream(lexer=lexer)
        parser = KplIteratorParser(stream)
        tree = parser.relationalExpression()
        s = self.visitLoOpExp(tree)
        return s

    def buildIterators(self, text):
        data = InputStream(text)
        lexer = KplIteratorLexer(data)
        stream = CommonTokenStream(lexer=lexer)
        parser = KplIteratorParser(stream)
        tree = parser.iterators()
        iterators, restrictions = self.visitIterators(tree)
        return iterators, restrictions, self.variables


# builder = KplIteratorBuilder({})
# s = builder.buildParameters('i=100, o, oo=1')
# print(s)
# s = builder.buildExpression('i+100+10*10+10')
# s = builder.buildExpression('i')
# print(s)
# s = builder.buildLogicalExpression('10>100&10>100|10<1000')
# print(s, s.evaluate())
# iterators, restrictions, variables = builder.buildIterators('1<o<10, 10+o<a<100,20, 10<p<100,i<100')
# print(iterators)
# print(restrictions)
# print(variables)

File: src/KPLinguaPreprocessing/KplIteratorBuilder.cs
using Antlr4.Runtime;
using KPLinguaPreprocessing.Grammar;
using KPLinguaPreprocessing.Models;
using System.Collections.Generic;
using static KPLinguaPreprocessing.Grammar.KplIteratorParser;

namespace KPLinguaPreprocessing
{
    public class KplIteratorBuilder : KplIteratorBaseVisitor<object>
    {
        private Dictionary<string, Variable> variables;

        public KplIteratorBuilder(Dictionary<string, Variable> variables)
        {
            this.variables = variables;
        }

        public (List<Iterator>, Base) VisitIterators(KplIteratorParser.IteratorsContext ctx)
        {
            var iterators = new List<Iterator>();
            foreach (var child in ctx.children)
            {
                var iterator = Visit(child) as Iterator;
                if (iterator != null)
                {
                    iterators.Add(iterator);
                }
            }
            Base restrictions = new RelationalExpressionTrue();
            if (ctx.restrictions != null)
            {
                restrictions = Visit(ctx.restrictions) as Base;
            }
            return (iterators, restrictions);
        }

        public override object VisitIterator(KplIteratorParser.IteratorContext ctx)
        {
            string leftInq = ctx.leftInq.Text == Sign.LessOrEqual ? Sign.LessOrEqual : Sign.Less;
            string rightInq = ctx.rightInq.Text == Sign.LessOrEqual ? Sign.LessOrEqual : Sign.Less;
            var e1 = Visit(ctx.left) as Base;
            var e2 = Visit(ctx.right) as Base;
            var identifier = ctx.Identifier().GetText();
            int increment = 1;
            if (ctx.increment != null)
            {
                increment = int.Parse(ctx.increment.Text);
            }

            if (!variables.TryGetValue(identifier, out var variable))
            {
                variable = new Variable(identifier);
                variables[identifier] = variable;
            }

            return new Iterator(variable, e1, leftInq, e2, rightInq, increment);
        }

        public override object VisitArOpExp(KplIteratorParser.ArOpExpContext ctx)
        {
            if (ctx.op != null)
            {
                string operation = ctx.op.Text;
                var e1 = Visit(ctx.left) as Base;
                var e2 = Visit(ctx.right) as Base;
                return new ArithmeticExpression(e1, operation, e2);
            }
            else if (ctx.Number() != null)
            {
                return new Number(int.Parse(ctx.GetText()));
            }
            else if (ctx.identifierOp != null)
            {
                return Visit(ctx.identifierOp);
            }
            return null;
        }

        public override object VisitLoOpExp(KplIteratorParser.LoOpExpContext ctx)
        {
            if (ctx.op != null)
            {
                var op = ctx.op.Text;
                var e1 = Visit(ctx.left) as Base;
                var e2 = Visit(ctx.right) as Base;
                return new RelationalExpression(e1, op, e2);
            }
            else
            {
                return Visit(ctx.lo);
            }
        }

        public override object VisitLogicalExpression(KplIteratorParser.LogicalExpressionContext ctx)
        {
            var ex1 = Visit(ctx.left) as Base;
            var op = Visit(ctx.lo);
            var ex2 = Visit(ctx.right) as Base;
            return new LogicalExpression(ex1, op.ToString(), ex2);
        }

        public override object VisitLogicalOperatorName(KplIteratorParser.LogicalOperatorNameContext ctx)
        {
            return LogicalOperators.GetOperator(ctx.GetText());
        }

        public override object VisitIdentifier(KplIteratorParser.IdentifierContext ctx)
        {
            var identifier = ctx.Identifier().GetText();
            if (!variables.TryGetValue(identifier, out var variable))
            {
                variable = new Variable(identifier);
                variables[identifier] = variable;
            }
            return variable;
        }

        public override object VisitParameters(KplIteratorParser.ParametersContext ctx)
        {
            var variables = new List<Variable>();
            foreach (var child in ctx.children)
            {
                var variable = Visit(child) as Variable;
                if (variable != null)
                {
                    variables.Add(variable);
                }
            }
            return variables;
        }

        public override object VisitParameter(KplIteratorParser.ParameterContext ctx)
        {
            var identifier = ctx.Identifier().GetText();
            int number = ctx.Number() != null ? int.Parse(ctx.Number().GetText()) : 0;
            return new Variable(identifier, number);
        }

        public List<Variable> BuildParameters(string text)
        {
            var data = new AntlrInputStream(text);
            var lexer = new KplIteratorLexer(data);
            var stream = new CommonTokenStream(lexer);
            var parser = new KplIteratorParser(stream);
            var tree = parser.parameters();
            return VisitParameters(tree) as List<Variable>;
        }

        public Base BuildExpression(string text)
        {
            var data = new AntlrInputStream(text);
            var lexer = new KplIteratorLexer(data);
            var stream = new CommonTokenStream(lexer);
            var parser = new KplIteratorParser(stream);
            var tree = parser.arithmeticExpression() as ArOpExpContext;
            return VisitArOpExp(tree) as Base;
        }

        //public Base BuildLogicalExpression(string text)
        //{
        //    var data = new AntlrInputStream(text);
        //    var lexer = new KplIteratorLexer(data);
        //    var stream = new CommonTokenStream(lexer);
        //    var parser = new KplIteratorParser(stream);
        //    KplIteratorParser.RelationalExpressionContext tree = parser.relationalExpression();
        //    return VisitLoOpExp(tree) as Base;
        //}

        public (List<Iterator>, Base, Dictionary<string, Variable>) BuildIterators(string text)
        {
            var data = new AntlrInputStream(text);
            var lexer = new KplIteratorLexer(data);
            var stream = new CommonTokenStream(lexer);
            var parser = new KplIteratorParser(stream);
            var tree = parser.iterators();
            var (iterators, restrictions) = VisitIterators(tree);
            return (iterators, restrictions, variables);
        }
    }
}